src/utils/open-browser-cra/index.js

```diff
@@ -0,0 +1,146 @@
+    action = Actions.NONE
+  } else {
+    action = Actions.BROWSER
+  }
+  return { action, value, args }
+}
+
+const executeNodeScript = function (scriptPath, url) {
+  const extraArgs = process.argv.slice(2)
+  const child = spawn(process.execPath, [scriptPath, ...extraArgs, url], {
+    stdio: 'inherit',
+  })
+
+  child.on('close', (code) => {
+    if (code !== 0) {
+      console.log()
+      console.log(chalk.red('The script specified as BROWSER environment variable failed.'))
+      console.log(`${chalk.cyan(scriptPath)} exited with code ${code} .`)
+      console.log()
+    }
+  })
+
+  return true
+}
+
+const startBrowserProcess = function (browser, url, args) {
+  // If we're on OS X, the user hasn't specifically
+  // requested a different browser, we can try opening
+  // Chrome with AppleScript. This lets us reuse an
+  // existing tab when possible instead of creating a new one.
+  const shouldTryOpenChromiumWithAppleScript =
+    process.platform === 'darwin' && (typeof browser !== 'string' || browser === OSX_CHROME)
+
+  if (shouldTryOpenChromiumWithAppleScript) {
+    // Will use the first open browser found from list
+    const supportedChromiumBrowsers = [
+      'Google Chrome Canary',
+      'Google Chrome',
+      'Microsoft Edge',
+      'Brave Browser',
+      'Vivaldi',
+      'Chromium',
+    ]
+
+    for (const chromiumBrowser of supportedChromiumBrowsers) {
+      try {
+        // Try our best to reuse existing tab
+        // on OSX Chromium-based browser with AppleScript
+        execSync(`ps cax | grep "${chromiumBrowser}"`)
+        execSync(`osascript ${join(__dirname, 'openChrome.applescript')} "${encodeURI(url)}" "${chromiumBrowser}"`, {
+          cwd: __dirname,
+          stdio: 'ignore',
+        })
+        return true
+      } catch (_) {
+        // Ignore errors.
+      }
+    }
+  }
+
+  // Another special case: on OS X, check if BROWSER has been set to "open".
+  // In this case, instead of passing `open` to `opn` (which won't work),
+  // just ignore it (thus ensuring the intended behavior, i.e. opening the system browser):
+  // https://github.com/facebook/create-react-app/pull/1690#issuecomment-283518768
+  if (process.platform === 'darwin' && browser === 'open') {
+    browser = undefined
+  }
+
+  // If there are arguments, they must be passed as array with the browser
+  if (typeof browser === 'string' && args.length !== 0) {
+    browser = [browser].concat(args)
+  }
+
+  // Fallback to open
+  // (It will always open new tab)
+  try {
+    const options = { app: browser, wait: false, url: true }
+
+    // Prevent `unhandledRejection` error.
+    open(url, options).catch(() => {})
+    return true
+  } catch (_) {
+    return false
+  }
+}
+
+/**
+ * Reads the BROWSER environment variable and decides what to do with it. Returns
+ * true if it opened a browser or ran a node.js script, otherwise false.
+ */
+const openBrowser = function (url) {
+  const { action, value, args } = getBrowserEnv()
```

Comment:
It seems to me it might be possible to merge `getBrowserEnv()` and `openBrowser()` and remove the intermediary `Actions` enum. I.e. replace the `case` below by the conditions tested in `getBrowserEnv()` directly.